ipa-core/src/ff/boolean_array.rs

```diff
@@ -405,20 +478,84 @@ macro_rules! boolean_array_impl {
                 type Input = Boolean;
 
                 fn expand(v: &Boolean) -> Self {
-                    let mut result = <$name>::ZERO;
+                    let mut result = <$name as SharedValue>::ZERO;
                     for i in 0..usize::try_from(<$name>::BITS).unwrap() {
-                        result.set(i, *v);
+                        result.0.set(i, bool::from(*v));
                     }
                     result
                 }
             }
 
+            impl TryFrom<Vec<Boolean>> for $name {
+                type Error = ();
+                fn try_from(value: Vec<Boolean>) -> Result<Self, Self::Error> {
+                    if value.len() == $bits {
+                        Ok(value.into_iter().collect::<Self>())
+                    } else {
+                        Err(())
+                    }
+                }
+            }
+
+            impl ArrayBuild for $name {
+                type Input = Boolean;
+                type Builder = BooleanArrayBuilder<$name>;
+
+                fn builder() -> Self::Builder {
+                    BooleanArrayBuilder::new()
+                }
+            }
+
+            impl SharedValueArray<Boolean> for $name {
+                const ZERO_ARRAY: Self = <$name as SharedValue>::ZERO;
+
+                fn from_fn<F: FnMut(usize) -> Boolean>(mut f: F) -> Self {
```

Comment:
why do we need this function?

Reply:
For `StdArray` the implementation uses https://doc.rust-lang.org/std/array/fn.from_fn.html to avoid zero-initializing the array. Otherwise, it's just a different way of writing a for loop to fill the array.

It does appear this is unused in this PR (probably due to either revision of the changes or splitting things into multiple PRs). The function is used in my branch for vectorizing share conversion.